### ISSUE TITLE
No autocompletion when using Carthage

### ISSUE THREAD

**Issue by ryanbaldwin** [Quick/Quick#633]
## So I'm stumped. 

I'm using Carthage to bring in Quick and Nimble into my project. I'm also using Carthage to bring in some of our other internal frameworks. Our internal frameworks are great, however, the Quick and Nimble frameworks used in my tests receive no autocompletion love, and odd editor parsing/indexing behaviour. Everything will build fine, but syntax highlighting is haphazard, there's no autocompletion or code hints, etc. Here's some pics:

<img width="534" alt="screen shot 2016-11-04 at 9 15 45 am" src="https://cloud.githubusercontent.com/assets/955086/20006745/5fa3536e-a26f-11e6-8c40-6dd470137b8c.png">

<img width="253" alt="screen shot 2016-11-04 at 12 17 58 am" src="https://cloud.githubusercontent.com/assets/955086/20006763/737d862a-a26f-11e6-8b5c-862987183fb1.png">

I should note I'm only experiencing this with Quick and Nimble, but not the 1 or 2 other Carthage frameworks I'm depending on.

## Quick Setup
I'm on an mid/early 2015 MBP, running OS X 10.11.6, running Xcode 8.1

## Here's what I've done
1. I created a `Cartfile.private`
    > \# Frameworks for testing that are far better than Apple's crappy XCTest
    > github "Quick/Quick"
    > github "Quick/Nimble"
2. I do a `carthage update --platform iOS`, things are pulled, cities built, people fall in love.
3. I drag the two frameworks from the `Carthage/Build/iOS` folder and into the `breatheAgainTests` target's `Build Phases -> Link Binary With Libraries`
4. I add the two frameworks to the `breatheAgainTests` target's `Build Phases -> Copy Files`
    <img width="989" alt="screen shot 2016-11-04 at 9 26 38 am" src="https://cloud.githubusercontent.com/assets/955086/20007035/d48bbd3c-a270-11e6-87e4-4fee0c0916a6.png">

### Still not seeing consistent highlighting or parsing
After the above and still not seeing consistent highlighting or parsing, I decided to try the following. Between each of the steps taken I did a `Product -> Clean`, `Product -> Clean Build Folder`, and deleted the `~/Library/Developer/Xcode/DerivedData` directory.

- On all targets I set `Apple LLVM 8.0 - Language - Modules` -> `Enable Modules (C and Objective-C)` to `No`
- In the `breatheAgain` project -> `PROJECT breatheAgain` (note: the actual project, not a target) -> `Search Paths` -> `Framework Search Paths` I added `$(PROJECT_DIR)/../Carthage` as `recursive`
- I actually upgraded to Xcode 8.1 hoping blindly it would magically resolve this. I then blew away my Carthage directory and regenerated all the frameworks. Still no dice.

### A plea for help
Has anybody experienced this level of stubbornness with using Frameworks and not cocoa pods? I'm kinda shocked by how difficult this has been. Am I missing something in my setup? Do the apple gods hate me? Asking for a friend.

Any help would be _yuge_. I can't post a link to the project as it's, unfortunately, for the organization, but I will try to accommodate all and any requests for further information. 

Much love!

**Comment by istx25** [comment]
Hi there! Thanks for reaching out, @ryanbaldwin. Happy to help. 😺 I believe the problem you were experiencing has to do with your project setup so here's a few personalised resources:

I setup a fresh Swift 3 Xcode project with Quick and Nimble using Carthage. You can download the zip file [here](https://mega.nz/#!6lomwKhT!KZ6GxQfvlSbnAbElY62_gCUVpG7iOi3uZIcbPiZhMkU). Feel free to look around.

I recorded a _quick_ :smirk: video going through the process and explaining each step. You can download the video by going [here](https://mega.nz/#!ehwBGagA!nfSXGcWRZPSym7FU3PG44YQkJA_5PDYfmXJuPa-H03c).

tldr? 

1. Add the Quick and Nimble frameworks in the navigator area.
    - Ensure "Copy items if needed" is unchecked. We want to reference the frameworks from their original spot in the Carthage directory.
    - Recommended: Put them in a group named "Frameworks" and [put the group outside of the iOS and test targets](https://cloud.githubusercontent.com/assets/6799989/20016822/691e0f04-a2c1-11e6-861a-6894cde00def.png).
2. Navigate to "Build Phases" on the test target.
   - Create a new Run script (press the plus button on the top left).
   - Put `/usr/local/bin/carthage copy-frameworks` in the main section.
   - Add `$(SRCROOT)/Carthage/Build/iOS/Nimble.framework` in the "Input Files" section.
   - Add `$(SRCROOT)/Carthage/Build/iOS/Quick.framework` in the "Input Files" section.
3. Open your test file and import Quick and Nimble.
   - `import Quick` and `import Nimble`

And then you're good to go!

Let me know if you have any other problems or follow-up questions and I'd be happy to help you some more. 😄 ✨ ❤️

**Comment by ryanbaldwin** [comment]
@istx25 - You're right about XCTest. I didn't mean to sound snarky and, truth be told, I'd forgot that's what was actually in the file. You're right that XCTest supports a good foundation for others to build great tooling and technologies, and it's not something I intended to crap all over. It's been a long couple days and I think my frustrations leaked out a bit. 

Everybody has done great work here, and I super appreciate what y'all have done, and the incredible support you gave me in the above post (which I'll be trying now). You're right - be thankful and grateful instead of a soon to be cold, shivering, embittered canuck. ;)

😍

**Comment by istx25** [comment]
Totally understandable! I've been there and I'm pretty sure it's safe to say most, if not all of us have been. The "I wish I could chuck my Mac out the window" moments but you know you won't because your mac's worth too many thousands of dollars and you don't have enough to replace it. 😜

Anyways, let me know if the video + sample project is helpful. If that doesn't resolve your problem, comment back and we can try something else! I'm also Canadian btw so it's totally normal to be cold and shivering in Canada. 😉

**Comment by ryanbaldwin** [comment]
Quick update @istx25  - This is now working. Most of what I had in my project targets already aligned with what you had. 

There was one major difference, however, and I don't really understand why it would matter: I have a workspace at a level above my project (naturally), and my Carthage stuff was at the workspace level, not at the project level. 

I didn't think this would actually be a big deal so long as my Framework Search Path was set appropriately (`$(SRCROOT)/../Carthage/**`) but, apparently, it _is_ a big deal... but only within the unit test target. So very strange. Perhaps there's some kind of internal Xcode thing where assumptions are made, or how things are resolved in test targets or something... I'm not sure.

In any case, I moved my Cartfile* down one deeper, into the project directory, and ran a `carthage update --platform iOS` after setting up the various project stuff correctly. And it all worked just wonderfully.

A huge thank you for all the help you provided. It was, quite honestly, some of the best help I've ever received from the OSS community. You really went the extra distance. I wish I could give you some gold.

Cheers!

**Comment by istx25** [comment]
That's great to hear. I'm glad I was helpful. 💛 I'm here to help people and build the community so no worries. If you'd really like to _give me some gold_, buy me a coffee one day! ☕️ In the meantime, I'll look into if I can replicate the problem with the Xcode Workspace. I'm not aware of anything that would cause that but we can see if we can fix it (or get Apple to)! 🚀

**Comment by istx25** [comment]
Hit me up on Twitter or Telegram if you'd like. My username is the same as here. Would be nice to make a new connection. 💃

**Comment by Ronaldoh1** [comment]
@istx25 thank you for post. I was missing import of quick and nimble in my test class. My auto complete is working again. After hour and hours of pain ...it's working! are you an Xcode god? 

*send digital hug* dude thank so much!

**Comment by istx25** [comment]
Thank you for taking the time to comment something nice. ❤️ 
"Are you an Xcode god?" made my day. 😜

Glad I could help. 🙂

**Comment by Ronaldoh1** [comment]
@istx25 haha your welcome
